lib/lib_i2c/arduino-i2c-sgp41/src/SensirionI2CSgp41.cpp

```diff
@@ -0,0 +1,184 @@
+ *   this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+ * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include "SensirionI2CSgp41.h"
+#include "Arduino.h"
+#include "SensirionCore.h"
+#include <Wire.h>
+
+#define SGP41_I2C_ADDRESS 0x59
+
+SensirionI2CSgp41::SensirionI2CSgp41() {
+}
+
+void SensirionI2CSgp41::begin(TwoWire& i2cBus) {
+    _i2cBus = &i2cBus;
+}
+
+uint16_t SensirionI2CSgp41::executeConditioning(uint16_t defaultRh,
+                                                uint16_t defaultT,
+                                                uint16_t& srawVoc) {
+    uint16_t error;
+    uint8_t buffer[8];
+    SensirionI2CTxFrame txFrame =
+        SensirionI2CTxFrame::createWithUInt16Command(0x2612, buffer, 8);
+
+    error = txFrame.addUInt16(defaultRh);
+    error |= txFrame.addUInt16(defaultT);
+    if (error) {
+        return error;
+    }
+
+    error = SensirionI2CCommunication::sendFrame(SGP41_I2C_ADDRESS, txFrame,
+                                                 *_i2cBus);
+    if (error) {
+        return error;
+    }
+
+    delay(50);
+
+    SensirionI2CRxFrame rxFrame(buffer, 8);
+    error = SensirionI2CCommunication::receiveFrame(SGP41_I2C_ADDRESS, 3,
+                                                    rxFrame, *_i2cBus);
+    if (error) {
+        return error;
+    }
+
+    error |= rxFrame.getUInt16(srawVoc);
+    return error;
+}
+
+uint16_t SensirionI2CSgp41::measureRawSignals(uint16_t relativeHumidity,
+                                              uint16_t temperature,
+                                              uint16_t& srawVoc,
+                                              uint16_t& srawNox) {
+    uint16_t error;
+    uint8_t buffer[8];
+    SensirionI2CTxFrame txFrame =
+        SensirionI2CTxFrame::createWithUInt16Command(0x2619, buffer, 8);
+
+    error = txFrame.addUInt16(relativeHumidity);
+    error |= txFrame.addUInt16(temperature);
+    if (error) {
+        return error;
+    }
+
+    error = SensirionI2CCommunication::sendFrame(SGP41_I2C_ADDRESS, txFrame,
+                                                 *_i2cBus);
+    if (error) {
+        return error;
+    }
+
+    delay(50);
```

Comment:
I understand that it's supposed to be the official library. But such `delay()` are very problematic with Tasmota because they freeze evevery thing and makes the device unresponsive. It's considered a very bad practice in Tasmota to have any `delay()` for more than 5 ms.

Could it be refactored to make the read 50ms later in Tasmota event loop?

Reply:
Ahh, that makes sense! Thanks for the explanation. I'll see if I can get these fixed up